### PR TITLE
Fix remote branch checkout

### PR DESCRIPTION
**Pull request by cmurphy**
Currently, the user can specify a branch other than master to make
changes to, and modulesync will attempt to pull updates from that
branch before starting work. This is a problem because if the branch
does not exist remotely, the pull will fail.

This patch proposes the following workflow:

- If the repo is being cloned for the first time, check out the
  user-specified branch locally based on local master
- If the repo exists locally and the user-specified branch exists
  remotely, check out that branch locally and pull any changes from the
  remote branch before starting work.
- If the repo exists locally and the branch does not exist remotely,
  create a new branch based on origin/master.

### CODE REVIEW

**Comment by cmurphy** [comment]
Ping @raphink and @razorsedge

File: lib/modulesync/git.rb [hunk 1]
```diff
@@ -21,9 +21,15 @@ def self.pull(git_base, name, branch, opts)
       else
```
Comment (razorsedge) [inline lib/modulesync/git.rb]: You still need the following at line 19 to deal with a freshly cloned repo:
 puts "Switching to branch #{branch}"
 repo.checkout(branch)

**Comment by razorsedge** [comment]
@cmurphy It *looks* like is should work, but I have to brush up on the docs and runs some tests to confirm.

File: lib/modulesync/git.rb (hunk 1 above)
Comment (cmurphy) [inline lib/modulesync/git.rb]: Good point.

**Comment by cmurphy** [comment]
@razorsedge updated to deal with a freshly cloned repo.

**Comment by cmurphy** [comment]
@razorsedge got any time to check this for me?

**Comment by razorsedge** [comment]
@cmurphy Sorry for the delays.  I've ben a bit preoccupied preparing for my talk at [Puppet Camp Phoenix](https://puppetlabs.com/events/puppet-camp-phoenix).

I made an initial attempt at testing this only to discover I had built 7fd0cc8 instead.  Which was confusing as I think it was working.  I need to test some commits against my testing repo to make sure.

**Comment by razorsedge** [comment]
No joy.

```
[mike@idiocy modulesync_configs] (master *+$%)$ msync update -m'Test of bcafbba modulesync gem after --noop.'
Syncing test
Overriding any local changes to repositories in ./modules
Switching to branch develop
/home/mike/.gem/ruby/1.9.1/gems/git-1.2.8/lib/git/lib.rb:764:in `command': git pull 'origin' 'develop'  2>&1:From github.com:razorsedge/test (Git::GitExecuteError)
 * branch            develop    -> FETCH_HEAD
error: Untracked working tree file 'spec/spec.opts' would be overwritten by merge.  Aborting
	from /home/mike/.gem/ruby/1.9.1/gems/git-1.2.8/lib/git/lib.rb:630:in `pull'
	from /home/mike/.gem/ruby/1.9.1/gems/git-1.2.8/lib/git/base.rb:370:in `pull'
	from /home/mike/.gem/ruby/1.9.1/gems/modulesync-0.2.1/lib/modulesync/git.rb:35:in `pull'
	from /home/mike/.gem/ruby/1.9.1/gems/modulesync-0.2.1/lib/modulesync.rb:64:in `block in run'
	from /home/mike/.gem/ruby/1.9.1/gems/modulesync-0.2.1/lib/modulesync.rb:57:in `each'
	from /home/mike/.gem/ruby/1.9.1/gems/modulesync-0.2.1/lib/modulesync.rb:57:in `run'
	from /home/mike/.gem/ruby/1.9.1/gems/modulesync-0.2.1/bin/msync:8:in `<top (required)>'
	from /home/mike/.gem/ruby/1.9.1/bin/msync:23:in `load'
	from /home/mike/.gem/ruby/1.9.1/bin/msync:23:in `<main>'
```

**Comment by razorsedge** [comment]
The problem is basically, which of these [documented]https://github.com/schacon/ruby-git/blob/master/README.md() commands will happen? 
```ruby
     g.branch('new_branch').checkout
     g.branch('existing_branch').checkout
```

Here is what I see after running `msync update --noop` twice (once to create the local checkout and the second time for the error above):
```
[mike@idiocy modulesync_configs] (master *+$%)$ cd modules/test
[mike@idiocy test] (develop %)$ git branch -av
* develop                2303c91 Merge branch 'release/1.1.0'
  master                 2303c91 Merge branch 'release/1.1.0'
  remotes/origin/HEAD    -> origin/master
  remotes/origin/develop c1d4c32 Commit message 5
  remotes/origin/master  2303c91 Merge branch 'release/1.1.0'
```
Notice how develop has the same commit as master, but remotes/origin/develop and remotes/origin/master differ?  What is happening is that the repo is cloned and since the develop branch does not yet exist locally, it is checked out *from master* (`git checkout -b develop`).  This is why I had to use `repo.checkout(branch)`.

**Comment by razorsedge** [comment]
Change your PR to add:
```diff
[mike@idiocy modulesync]$ diff -u git.rb-orig git.rb
--- git.rb-orig	2015-02-20 14:27:20.630472872 -0700
+++ git.rb	2015-02-22 00:13:04.968379521 -0700
@@ -7,7 +7,7 @@
     def self.switch_branch(repo, branch)
       unless repo.branch.name == branch
         puts "Switching to branch #{branch}"
-        repo.branch(branch).checkout
+        repo.checkout(branch)
       end
     end
 
@@ -32,6 +32,7 @@
         repo.reset_hard
         if repo.branches.remote.collect { |r| r.name }.include?(branch)
           switch_branch(repo, branch)
+          puts "Pulling updates from origin/#{branch}"
           repo.pull('origin', branch)
         else # git checkout -b branch origin/master
           repo.checkout('origin/master')
@@ -44,7 +45,7 @@
     # Git add/rm, git commit, git push
     def self.update(name, files, message, branch)
       repo = ::Git.open("#{PROJ_ROOT}/#{name}")
-      repo.branch(branch).checkout
+      switch_branch(repo, branch)
       files.each do |file|
         if repo.status.deleted.include?(file)
           repo.remove(file)
@@ -77,7 +78,7 @@
       puts "Using no-op. Files in #{name} may be changed but will not be committed."
 
       repo = ::Git.open("#{PROJ_ROOT}/#{name}")
-      repo.branch(branch).checkout
+      switch_branch(repo, branch)
 
       puts "Files changed: "
       repo.diff('HEAD', '--').each do |diff|
```

**Comment by cmurphy** [comment]
I'm not able to reproduce your error:

```
$ git branch -av
* master                   cf6457f Fix spec tests
  remotes/origin/HEAD      -> origin/master
  remotes/origin/master    cf6457f Fix spec tests
  remotes/origin/newbranch 2fbcb6d test
```
notice remote branch that doesn't exist locally ^

```
$ cd -
/Users/colleen/dev/modulesync_configs
$ ../modulesync/bin/msync update --noop
No config file under ./config_defaults.yml found, using default values
Syncing puppetlabs-rabbitmq
Overriding any local changes to repositories in ./modules
Switching to branch newbranch
No config file under ./modules/puppetlabs-rabbitmq/.sync.yml found, using default values
Using no-op. Files in puppetlabs-rabbitmq may be changed but will not be committed.
Files changed:
Files added:
testfile2


--------------------------------
$ cd -
/Users/colleen/dev/modulesync_configs/modules/puppetlabs-rabbitmq
$ git branch -av
  master                   cf6457f Fix spec tests
* newbranch                2fbcb6d test
  remotes/origin/HEAD      -> origin/master
  remotes/origin/master    cf6457f Fix spec tests
  remotes/origin/newbranch 2fbcb6d test
$ cd -
/Users/colleen/dev/modulesync_configs
$ ../modulesync/bin/msync update --noop
No config file under ./config_defaults.yml found, using default values
Syncing puppetlabs-rabbitmq
Overriding any local changes to repositories in ./modules
Switching to branch newbranch
No config file under ./modules/puppetlabs-rabbitmq/.sync.yml found, using default values
Using no-op. Files in puppetlabs-rabbitmq may be changed but will not be committed.
Files changed:
Files added:
testfile2


--------------------------------
$ cd -
/Users/colleen/dev/modulesync_configs/modules/puppetlabs-rabbitmq
$ git branch -av
  master                   cf6457f Fix spec tests
* newbranch                2fbcb6d test
  remotes/origin/HEAD      -> origin/master
  remotes/origin/master    cf6457f Fix spec tests
  remotes/origin/newbranch 2fbcb6d test
```

The trouble with your approach is that if the branch exists neither locally nor remotely, `repo.checkout(branch)` will fail. It needs to be able to create the branch if needed. `repo.branch(branch).checkout` *should* have the correct behavior in both cases - I'm unclear why it's not the behavior you're experiencing :(

**Comment by razorsedge** [comment]
All I can think of, then is environment:
```
[mike@idiocy modulesync_configs] (master *$%)$ which ruby
/opt/rh/ruby193/root/usr/bin/ruby
[mike@idiocy modulesync_configs] (master *$%)$ ruby -v
ruby 1.9.3p484 (2013-11-22) [x86_64-linux]
[mike@idiocy modulesync_configs] (master *$%)$ gem list|grep git
git (1.2.8)
[mike@idiocy modulesync_configs] (master *$%)$ rpm -q git
git-1.7.1-3.el6_4.1.x86_64
[mike@idiocy modulesync_configs] (master *$%)$ cat /etc/redhat-release 
CentOS release 6.6 (Final)
```

**Comment by cmurphy** [comment]
Aha!

I'm using git version 2. Switching to git version 1 produces the problem you've described.

I'll keep poking at it to get it to work with git 1.

**Comment by cmurphy** [comment]
Okay, pushed a new change - this should hopefully now be able to distinguish between when a local branch exists that can be updated via pull and when a remote branch needs to be checked out.

**Comment by razorsedge** [comment]
Yay!

Except now I am getting a new error:
```
[mike@idiocy modulesync_configs] (master *$%)$ rm -rf modules/test/
[mike@idiocy modulesync_configs] (master *$%)$ msync update --noop
Syncing test
Cloning repository fresh
Cloning from git@github.com:razorsedge/test.git
Creating local branch develop from origin/develop
No config file under ./modules/test/.sync.yml found, using default values
Using no-op. Files in test may be changed but will not be committed.
Files changed: 
Files added: 

--------------------------------
[mike@idiocy modulesync_configs] (master *$%)$ msync update --noop
Syncing test
Overriding any local changes to repositories in ./modules
Switching to branch develop
/home/mike/.gem/ruby/1.9.1/gems/git-1.2.9.1/lib/git/lib.rb:917:in `command': git '--git-dir=/home/mike/git/razorsedge/modulesync_configs/modules/test/.git' '--work-tree=/home/mike/git/razorsedge/modulesync_configs/modules/test' pull 'origin' 'develop'  2>&1:fatal: /usr/libexec/git-core/git-pull cannot be used without a working tree. (Git::GitExecuteError)
	from /home/mike/.gem/ruby/1.9.1/gems/git-1.2.9.1/lib/git/lib.rb:732:in `pull'
	from /home/mike/.gem/ruby/1.9.1/gems/git-1.2.9.1/lib/git/base.rb:351:in `pull'
	from /home/mike/.gem/ruby/1.9.1/gems/modulesync-0.2.2/lib/modulesync/git.rb:49:in `pull'
	from /home/mike/.gem/ruby/1.9.1/gems/modulesync-0.2.2/lib/modulesync.rb:64:in `block in run'
	from /home/mike/.gem/ruby/1.9.1/gems/modulesync-0.2.2/lib/modulesync.rb:57:in `each'
	from /home/mike/.gem/ruby/1.9.1/gems/modulesync-0.2.2/lib/modulesync.rb:57:in `run'
	from /home/mike/.gem/ruby/1.9.1/gems/modulesync-0.2.2/bin/msync:8:in `<top (required)>'
	from /home/mike/.gem/ruby/1.9.1/bin/msync:23:in `load'
	from /home/mike/.gem/ruby/1.9.1/bin/msync:23:in `<main>'
```

**Comment by cmurphy** [comment]
@razorsedge try now?

**Comment by razorsedge** [comment]
@cmurphy That appears to work for me.

**Comment by cmurphy** [comment]
\o/

**Comment by razorsedge** [comment]
Fixes #34.

**Pull request merged**